lib/kubernetes-deploy/kubernetes_resource/pod.rb

```diff
@@ -120,5 +140,67 @@ def display_logs
 
       @already_displayed = true
     end
+
+    class Container
+      STATUS_SCAFFOLD = {
+        "state" => {
+          "running" => {},
+          "waiting" => {},
+          "terminated" => {},
+        },
+        "lastState" => {
+          "terminated" => {}
+        }
+      }.freeze
+
+      attr_reader :name, :probe_location
+
+      def initialize(definition, init: false)
+        @init = init
+        @name = definition["name"]
+        @image = definition["image"]
+        @probe_location = definition.fetch("readinessProbe", {}).fetch("httpGet", {})["path"]
+        @status = STATUS_SCAFFOLD.dup
+      end
+
+      def doomed?
+        doom_reason.present?
+      end
+
+      def doom_reason
+        exit_code = @status['lastState']['terminated']['exitCode']
+        last_terminated_reason = @status["lastState"]["terminated"]["reason"]
+        limbo_reason = @status["state"]["waiting"]["reason"]
+
+        if last_terminated_reason == "ContainerCannotRun"
+          # ref: https://github.com/kubernetes/kubernetes/blob/562e721ece8a16e05c7e7d6bdd6334c910733ab2/pkg/kubelet/dockershim/docker_container.go#L353
+          "Failing to start (exit #{exit_code}): #{@status['lastState']['terminated']['message']}"
+        elsif limbo_reason == "CrashLoopBackOff"
+          "Crashing repeatedly (exit #{exit_code}). See logs for more information."
+        elsif %w(ImagePullBackOff ErrImagePull).include?(limbo_reason)
```

Comment:
> One concern I have here is that on large deployments a single container going into ImagePullBackOff will fail the entire deploy, when it could be a recoverable condition. 

Since the condition is `@pods.present? && @pods.all?(&:deploy_failed?)` (all?, not any?), it would only fail the deploy if they _all_ hit that same transient failure. You make a very good point though. It is easy enough to look for "not found". Could be a little fragile, but a test will fail if a future k8s version changes the message, so I'll do it 👍 .